src/symbolize/gimli/elf.rs

```diff
@@ -10,11 +15,150 @@ type Elf = object::elf::FileHeader32<NativeEndian>;
+        let object = Object::parse(&map)?;
+
+        // Try to locate an external debug file using the build ID.
+        let build_id = object.build_id().unwrap_or(&[]);
+        if !build_id.is_empty() {
+            fn hex(byte: u8) -> u8 {
+                if byte < 10 {
+                    b'0' + byte
+                } else {
+                    b'a' + byte - 10
+                }
+            }
+            let mut path = Vec::with_capacity(
+                BUILD_ID_PATH.len() + BUILD_ID_SUFFIX.len() + build_id.len() * 2 + 1,
+            );
+            path.extend(BUILD_ID_PATH);
+            path.push(hex(build_id[0] >> 4));
+            path.push(hex(build_id[0] & 0xf));
+            path.push(b'/');
+            for byte in &build_id[1..] {
+                path.push(hex(byte >> 4));
+                path.push(hex(byte & 0xf));
+            }
+            path.extend(BUILD_ID_SUFFIX);
+            if let Some(mapping) = Mapping::new_debug(Path::new(OsStr::from_bytes(&path)), None) {
+                return Some(mapping);
+            }
+        }
+
+        // Try to locate an external debug file using the GNU debug link section.
+        let tmp_stash = Stash::new();
+        if let Some(section) = object.section(&tmp_stash, ".gnu_debuglink") {
+            if let Some(len) = section.iter().position(|x| *x == 0) {
+                let filename = &section[..len];
+                let offset = (len + 1 + 3) & !3;
+                if let Some(crc_bytes) = section
+                    .get(offset..offset + 4)
+                    .and_then(|bytes| bytes.try_into().ok())
+                {
+                    let crc = u32::from_ne_bytes(crc_bytes);
+                    if let Some(path_debug) = locate_debuglink(path, filename) {
+                        if let Some(mapping) = Mapping::new_debug(&path_debug, Some(crc)) {
+                            return Some(mapping);
+                        }
+                    }
+                }
+            }
+        }
+
+        let stash = Stash::new();
+        let cx = Context::new(&stash, object)?;
+        Some(Mapping {
+            // Convert to 'static lifetimes since the symbols should
+            // only borrow `map` and `stash` and we're preserving them below.
+            cx: unsafe { core::mem::transmute::<Context<'_>, Context<'static>>(cx) },
+            _map: map,
+            _map_sup: None,
+            _stash: stash,
+        })
+    }
+
+    /// Load debuginfo from an external debug file.
+    fn new_debug(path: &Path, crc: Option<u32>) -> Option<Mapping> {
+        let map = super::mmap(path)?;
+        let object = Object::parse(&map)?;
+
+        if let Some(_crc) = crc {
+            // TODO: check crc
+        }
+
+        // Try to locate a supplementary object file.
+        let tmp_stash = Stash::new();
+        if let Some(section) = object.section(&tmp_stash, ".gnu_debugaltlink") {
```

Comment:
Oh deear, I can't seem to find documentation on this... Do you know of some good online docs to link to?

I'm curious about why this is handled here rather than above from the original object...

Reply:
I don't know of any documentation for the format of this. The only documentation I found is the dwz man page, but it doesn't give this detail. I based the code on what libbacktrace does, but gdb would have been a better source. And indeed, gdb does the path resolution differently so I'll have to fix that.

It is handled here because this is a section in the external debug file, not the original object.